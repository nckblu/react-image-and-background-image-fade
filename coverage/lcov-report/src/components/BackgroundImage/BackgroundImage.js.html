<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/Backgroundimage/BackgroundImage.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">src/components/Backgroundimage</a> BackgroundImage.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>18/18</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">95.83% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>23/24</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/2</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>18/18</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">19x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">19x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">19x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { Fragment } from "react";
import Loader from "../Loader";
import ImageLoader from "../ImageLoader";
import { Wrapper } from "./elements/Wrapper";
import PropTypes from "prop-types";
import defaults from "../../defaults";
import { getAspectRatioPercentage } from "../../util";
&nbsp;
export const BackgroundImage = ({
  src,
  width,
  height,
  transitionTime,
  renderLoader,
  disableLoader,
  useChild,
  children,
  element,
  lazyLoad,
  wrapperClassName,
  isResponsive,
  ...props
}) =&gt; (
  &lt;ImageLoader src={src} transitionTime={transitionTime} lazyLoad={lazyLoad}&gt;
    {({ hasLoaded, shouldShowLoader, hasFailed, src }) =&gt; {
      const backgroundImageStyle = { backgroundImage: `url("${src}")` };
      let style;
      let childElement;
      if (useChild) {
        /*
         * When in useChild mode we need to clone the child, append the style
         * and add relative positioning for the loader.
         */
        style =
          children &amp;&amp; children.props &amp;&amp; children.props.style
            ? { ...children.props.style }
            : {};
        style.position = "relative";
        hasLoaded ? (style = { ...style, ...backgroundImageStyle }) : style;
        childElement = React.cloneElement(children, { style });
      } else {
        /*
         * When not in useChild mode we need to create a new element based on
         * the element prop, apply the styles and include the current children
         */
        style = props.style ? <span class="branch-0 cbranch-no" title="branch not covered" >{ ...props.style } </span>: {};
        style.width = !isResponsive ? width : "100%";
        style.height = !isResponsive ? height : "auto";
        if (isResponsive) {
          style.paddingTop = getAspectRatioPercentage(width, height);
        }
&nbsp;
        hasLoaded ? (style = { ...style, ...backgroundImageStyle }) : style;
        childElement = React.createElement(
          element,
          { ...props, style },
          children
        );
      }
&nbsp;
      return (
        &lt;Wrapper
          width={width}
          height={height}
          isResponsive={isResponsive}
          className={wrapperClassName}
        &gt;
          {childElement}
          {shouldShowLoader &amp;&amp; !disableLoader &amp;&amp; (
            &lt;Fragment&gt;
              {renderLoader ? (
                renderLoader({ hasLoaded, hasFailed })
              ) : (
                &lt;Loader
                  isOnTop
                  hasLoaded={hasLoaded}
                  transitionTime={transitionTime}
                /&gt;
              )}
            &lt;/Fragment&gt;
          )}
        &lt;/Wrapper&gt;
      );
    }}
  &lt;/ImageLoader&gt;
);
&nbsp;
BackgroundImage.propTypes = {
  src: PropTypes.string.isRequired,
  width: PropTypes.string,
  height: PropTypes.string,
  transitionTime: PropTypes.string,
  renderLoader: PropTypes.func,
  disableLoader: PropTypes.bool,
  useChild: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  element: PropTypes.string,
  style: PropTypes.object,
  wrapperClassName: PropTypes.string,
  lazyLoad: PropTypes.bool,
  isResponsive: PropTypes.bool,
};
&nbsp;
BackgroundImage.defaultProps = {
  transitionTime: defaults.transitionTime,
  element: "div",
};
&nbsp;
export default BackgroundImage;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Tue Feb 12 2019 22:39:15 GMT+0000 (GMT)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
