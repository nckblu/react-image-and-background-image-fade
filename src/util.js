export const cssTimeToMs = cssTime => {
  const num = parseFloat(cssTime, 10);
  let unit = cssTime.match(/m?s/);
  let milliseconds;

  if (unit) {
    unit = unit[0];
  }

  switch (unit) {
    case "s":
      milliseconds = num * 1000;
      break;
    case "ms":
      milliseconds = num;
      break;
    default:
      milliseconds = 0;
      break;
  }

  return milliseconds;
};

export const getAspectRatioPercentage = (width, height) => {
  const widthNumber = parseInt(width, 10);
  const heightNumber = parseInt(height, 10);
  const ratio = (heightNumber / widthNumber) * 100;
  return `${ratio}%`;
};
