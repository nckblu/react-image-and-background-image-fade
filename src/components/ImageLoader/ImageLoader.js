import React from "react";
import PropTypes from "prop-types";
import preloader from "image-preloader";
import { cssTimeToMs } from "../../util";
import defaults from "../../defaults";
import VisibilitySensor from "react-visibility-sensor";

export class ImageLoader extends React.Component {
  static propTypes = {
    src: PropTypes.string.isRequired,
    transitionTime: PropTypes.string,
    lazyLoad: PropTypes.bool,
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
      PropTypes.func,
    ]),
  };

  state = {
    hasLoaded: false,
    hasFailed: false,
    shouldShowLoader: true,
    visibilitySensorIsActive: true,
  };

  componentDidMount() {
    if (!this.props.lazyLoad) {
      this.preloadImage();
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.src !== this.props.src) {
      if (!this.props.lazyLoad) {
        this.preloadImage();
      } else {
        this.setState({ visibilitySensorIsActive: true });
      }
    }
  }

  render() {
    const { src, children, lazyLoad } = this.props;
    const {
      hasLoaded,
      hasFailed,
      shouldShowLoader,
      visibilitySensorIsActive,
    } = this.state;
    /*
     * When using ImageLoader the children prop should be a function.
     * Render calls this function with the props it requires to
     * fade in an image.
     * See https://reactjs.org/docs/render-props.html
     */
    const childElement = children({
      hasLoaded,
      shouldShowLoader,
      hasFailed,
      src,
    });

    /*
     * Wrap element in react visibility sensor when lazyLoad is true
     */
    return lazyLoad ? (
      <VisibilitySensor
        active={visibilitySensorIsActive}
        onChange={this.handleVisibilityChange}
        partialVisibility
      >
        {childElement}
      </VisibilitySensor>
    ) : (
      childElement
    );
  }

  preloadImage() {
    const { src } = this.props;
    preloader
      .simplePreload(src)
      .then(this.handleImageLoaderLoaded)
      .catch(this.handleImageLoaderFailed);
  }

  /*
   * Once the image is loaded successfully this handler will be
   * called, this will change the `hasLoaded` attribute to true
   * then the `shouldShowLoader` attribute to false after
   * the transition has completed. The reason we do this on a
   * timeout is to ensure the loader is not hidden as soon as the
   * image has loaded resulting in a 'flash'
   */
  handleImageLoaderLoaded = () => {
    const { transitionTime } = this.props;
    const hideLoaderDelay = cssTimeToMs(transitionTime);
    this.setState({ hasLoaded: true }, () => {
      setTimeout(() => {
        this.setState({
          shouldShowLoader: false,
        });
      }, hideLoaderDelay);
    });
  };

  handleImageLoaderFailed = () => {
    this.setState({ hasFailed: true });
  };

  handleVisibilityChange = isVisible => {
    if (!isVisible) return;
    this.setState({ visibilitySensorIsActive: false });
    this.preloadImage();
  };
}

ImageLoader.defaultProps = {
  transitionTime: defaults.transitionTime,
};

export default ImageLoader;
