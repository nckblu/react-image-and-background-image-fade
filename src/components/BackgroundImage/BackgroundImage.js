import React, { Fragment } from "react";
import styled, { keyframes } from "styled-components";
import preloader from "image-preloader";
import Loader from "../Loader";
import { cssTimeToMs } from "../../util";
import ImageLoader from "../ImageLoader";
import { fadeIn } from "../../keyframes";
import { relative } from "path";

export const BackgroundImage = ({
  src,
  width,
  height,
  transitionTime,
  renderLoader,
  disablePlaceholder,
  useChild,
  children,
  ...props
}) => (
  <ImageLoader
    src={src}
    width={width}
    height={height}
    transitionTime={transitionTime}
    renderLoader={renderLoader}
    disablePlaceholder={disablePlaceholder}
  >
    {({
      hasLoaded,
      shouldShowLoader,
      hasFailed,
      src,
      transitionTime,
      disablePlaceholder,
      renderLoader,
    }) => {
      if (useChild) {
        // Preserve initial styles if the style prop is being used on the child element
        const initialStyles = children.props.style
          ? { ...children.props.style }
          : {};

        const styles = {
          position: relative, // Component requires relevant positioning to accomodate the loader
        };

        if (hasLoaded) {
          styles.backgroundImage = `url("${src}")`;
        }
        const newChild = React.cloneElement(children, { styles });
      } else {
        return <span />;
      }
    }}
  </ImageLoader>
);
BackgroundImage.defaultProps = {
  transitionTime: "0.3s",
};

export default BackgroundImage;

const Wrapper = styled.div`
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  position: relative;
`;

const Img = styled.img`
  width: 100%;
  height: 100%;
  animation: ${fadeIn} ${props => props.transitionTime} ease;
`;
