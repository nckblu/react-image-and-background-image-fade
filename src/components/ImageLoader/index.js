import ImageLoader from "./ImageLoader";

export default ImageLoader;
