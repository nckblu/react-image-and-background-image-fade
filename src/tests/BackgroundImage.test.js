import React from "react";
import BackgroundImage from "../components/BackgroundImage";
import renderer from "react-test-renderer";
import { simplePreloadMock } from "./__mocks__/ImageLoader";
import { mount } from "enzyme";
import { flushPromises } from "./util";

jest.mock("image-preloader");
const ImagePreloader = require("image-preloader");
ImagePreloader.simplePreload.mockImplementation(simplePreloadMock);

const renderLoader = () => <div className='Loading'>We loading up in here</div>;

describe("BackgroundImage", () => {
  const genericProps = {
    src: "me-drinking-pina-colada.jpg",
    width: "300px",
    height: "300px",
    wrapperClassName: "MaWrapperClass",
  };

  it("Matches snapshot of generic props", () => {
    const props = {
      ...genericProps,
    };
    const tree = renderer.create(<BackgroundImage {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should display loader while image is loading", () => {
    const props = {
      ...genericProps,
      renderLoader,
    };
    const wrapper = mount(<BackgroundImage {...props} />);
    expect(wrapper.find(".Loading").length).toEqual(1);
  });

  it("Shouldn't display loader while image is loading when disableLoader is true", () => {
    const props = {
      ...genericProps,
      renderLoader,
      disableLoader: true,
    };
    const wrapper = mount(<BackgroundImage {...props} />);
    expect(wrapper.find(".Loading").length).toEqual(0);
  });

  it("Should hide loader once loaded", async () => {
    const props = {
      ...genericProps,
      renderLoader,
    };
    jest.useFakeTimers();
    const wrapper = mount(<BackgroundImage {...props} />);
    /*
     * Here we need to use some poopery to ensure that
     * the promise is resolved, the component's update is
     * triggered, then to wait for the transition time to complete
     * then trigger the component update again so we can check
     * that the loader has gone.
     */
    await flushPromises();
    wrapper.update();
    jest.runAllTimers();
    wrapper.update();
    jest.useRealTimers();

    expect(wrapper.find(".Loading").length).toEqual(0);
  });

  it("Should display background image once loaded", async () => {
    const props = {
      ...genericProps,
      renderLoader,
    };

    const wrapper = mount(<BackgroundImage {...props} />);
    await flushPromises();
    wrapper.update();

    expect(
      wrapper
        .find("div.MaWrapperClass > div")
        .first()
        .props().style.backgroundImage
    ).toEqual(`url("${props.src}")`);
  });

  it("Uses the element prop correctly for the creation of the background image", async () => {
    const props = {
      ...genericProps,
      renderLoader,
      element: "p",
    };

    const wrapper = mount(<BackgroundImage {...props} />);
    await flushPromises();
    wrapper.update();
    expect(wrapper.find("div.MaWrapperClass > p").length).toEqual(1);
  });

  it("Attaches background-image CSS to child when using useChild prop", async () => {
    const props = {
      ...genericProps,
      renderLoader,
      useChild: true,
    };

    const wrapper = mount(
      <BackgroundImage {...props}>
        <div className='ImAChild' />
      </BackgroundImage>
    );
    await flushPromises();
    wrapper.update();
    expect(wrapper.find(".ImAChild").props().style.backgroundImage).toEqual(
      `url("${props.src}")`
    );
  });

  it("Preserves child style when using useChild prop", async () => {
    const props = {
      ...genericProps,
      renderLoader,
      useChild: true,
    };

    const childStyle = { fontSize: "20000px" };
    const wrapper = mount(
      <BackgroundImage {...props}>
        <div className='ImAChild' style={childStyle} />
      </BackgroundImage>
    );
    await flushPromises();
    wrapper.update();
    expect(wrapper.find(".ImAChild").props().style.fontSize).toEqual("20000px");
  });

  it("Sets paddingTop aspect ratio when using isResponsive prop", async () => {
    const props = {
      ...genericProps,
      renderLoader,
      isResponsive: true,
    };

    const wrapper = mount(<BackgroundImage {...props} />);
    await flushPromises();
    wrapper.update();
    expect(
      wrapper
        .find("div.MaWrapperClass > div")
        .first()
        .props().style.paddingTop
    ).toBeTruthy();
  });
});
