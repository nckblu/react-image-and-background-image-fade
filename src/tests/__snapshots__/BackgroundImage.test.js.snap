// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage Matches snapshot of generic props 1`] = `
<div
  className="MaWrapperClass sc-bwzfXH MWGHc"
  height="300px"
  width="300px"
>
  <div
    style={
      Object {
        "height": "300px",
        "width": "300px",
      }
    }
  />
  <div
    className="sc-bdVaJa iVGwIn"
  />
</div>
`;
